RTIC 2

korken89 opened the pull request.

korken89 committed:
RTIC v2: Initial commit

rtic-syntax is now part of RTIC repository

korken89 committed:
Old xtask test pass

korken89 committed:
Add check again

korken89 committed:
syntax: Remove parse settings struct

korken89 committed:
Min codegen

korken89 committed:
Added software task codegen back

korken89 committed:
Fix fences

korken89 committed:
Even more cleanup

korken89 committed:
First example builds again

korken89 committed:
More removal

korken89 committed:
Removed Priority, simplified lifetime handling

korken89 committed:
Break codegen for 0-prio async

korken89 committed:
Lifetime cleanup

korken89 committed:
Removed same prio spawn

perlindgren committed:
examples/async-task fixup

korken89 committed:
Main in main codegen

perlindgren committed:
export Cell removed, expmples updated

perlindgren committed:
examples/idle-wfi fixed

korken89 committed:
Cleanup weird locals in codegen

perlindgren committed:
examples/locals fixed

perlindgren committed:
examples/lock fixed

korken89 committed:
All codegen is now explicit

perlindgren committed:
examples/multiloc fixed

korken89 committed:
Fix locks, basepri writeback error

perlindgren committed:
more examples

korken89 committed:
Update UI tests, 1 failing that needs fixing

korken89 committed:
Fix failing UI test

korken89 committed:
Fix typos

korken89 committed:
Update semihosting

korken89 committed:
All examples pass with `cargo xtask --target all`

korken89 committed:
Support 0 prio tasks

korken89 committed:
Clippy fixes

korken89 committed:
Start CI, disable docs building

korken89 committed:
New executor design

korken89 committed:
More work on new spawn/executor

korken89 committed:
Async tasks can now take arguments at spawn again

perlindgren committed:
executor update for less unsafe and more clear

korken89 committed:
monotonic experiments

korken89 committed:
Fix so deny(missing_docs) work

korken89 committed:
Add rtic-timer (timerqueue + monotonic) and rtic-monotonics (systick-monotonic)

korken89 committed:
Does CI work again?

korken89 committed:
Add gitignore for book

korken89 committed:
Fixed systick monotonic

korken89 committed:
Optimize linked list popping so delete is not run everytime

korken89 committed:
Systick runs at 1 kHz

korken89 committed:
Remove unnecessary MaybeUninit

korken89 committed:
Fixes in MPSC linked list and dropper handling

AfoHT committed:
Make clippy and fmt happy

AfoHT committed:
Experiment with changelog enforcer per path

AfoHT committed:
CI: Changelog fix syntax

AfoHT committed:
CI: Changelog: s/timer/time/

AfoHT committed:
CI: Clippy for time, monotonics, channel

AfoHT committed:
waker registration somehow lost, back again

AfoHT committed:
CI: Don't let warnings get away

korken89 committed:
Remove parsing on `capacity`

korken89 committed:
Added examples for async crates + fixed codegen for non-Copy arguments

AfoHT commented on rtic-monotonics/Cargo.toml:
```diff
@@ -0,0 +1,13 @@
+[package]
+name = "rtic-monotonics"
+version = "0.1.0"
```
Both rtic-channel and rtic-monotonics start at 1.0.0 currently

Should we maybe do "v1.0.0-alpha.0" for all of these, and then when RTIC v2 reaches stable these hopefully also reached stable v1?

korken89 replied:
SGTM

perlindgren replied:
Is it really needed, nobody will rely on them being stable in any case?

AfoHT committed:
Add changelog templates

AfoHT committed:
CI: Add rtic-channel to Changelog, remove defunct changelog

AfoHT committed:
CI: Cargo fmt for channel, mono., time

AfoHT committed:
CI: Alphabetical sort of clippy jobs

AfoHT committed:
CI: Cleanup old syntax CI

AfoHT committed:
Book: Fix gitignore to exclude mdbook output

korken89 committed:
Use `Pin` in the linked lists

korken89 committed:
rtic-channel: Fix clippy lint

AfoHT committed:
CI: Check and tests for all crates

AfoHT committed:
CI: Check/build the docs

Still no publish or further steps

AfoHT committed:
Fix some references to cortex-m-rtic

korken89 committed:
rtic-channel: Add testing, fix bugs

korken89 committed:
Fix CI for rtic-channel

korken89 committed:
rtic-time: Make Send happy

perlindgren committed:
Book: Major rework for RTIC v2

korken89 committed:
Move common data structures to `rtic-common`

korken89 committed:
Add `rtic-arbiter`

AfoHT commented on book/en/src/preface.md:
```diff
@@ -10,29 +10,138 @@
-This is the documentation of v1.0.x of RTIC; for the documentation of version
+This is the documentation of v2.0.x (pre-release) of RTIC 2.
 
-* v0.5.x go [here](/0.5).
-* v0.4.x go [here](/0.4).
```
TODO: Bring back links and add docs for v1

korken89 committed:
Added testing to rtic-arbiter

AfoHT commented on book/en/src/by-example/app.md:
```diff
@@ -2,25 +2,31 @@
 
+## Structure and zero-cost concurrency
+
+An RTIC `app` is an executable system model for since-core applications, declaring a set of `local` and `shared` resources operated on by a set of `init`, `idle`, *hardware* and *software* tasks. In short the `init` task runs before any other task returning the set of `local` and `shared` resources. Tasks run preemptively based on their associated static priority, `idle` has the lowest priority ( ...
```
```suggestion
An RTIC `app` is an executable system model for single-core applications, declaring a set of `local` and `shared` resources operated on by a set of `init`, `idle`, *hardware* and *software* tasks. In short the `init` task runs before any other task returning the set of `local` and `shared` resources. Tasks run preemptively based on their associated static priority, `idle` has the lowest priority (and can be used for background work, and/or to put the system to sleep until woken by some event). Hardware tasks are bound to underlying hardware interrupts, while software tasks are scheduled by asynchronous executors (one for each software task priority). 
```

AfoHT committed:
CI: Add mdbook-mermaid

AfoHT committed:
Book: Enable mermaid for mdbook

AfoHT committed:
Book: Add mermaid files

korken89 committed:
rtic-channel: try_* APIs now error if Sender/Receiver does not exist

korken89 committed:
Fix spelling error

AfoHT commented on book/en/src/preface.md:
```diff
@@ -10,29 +10,160 @@
+[Timber]: https://timber-lang.org/
+[RTFM-SRP]: https://www.diva-portal.org/smash/get/diva2:1005680/FULLTEXT01.pdf
+[RTFM-core]: https://ltu.diva-portal.org/smash/get/diva2:1013248/FULLTEXT01.pdf
+[AbstractTimer]: https://ltu.diva-portal.org/smash/get/diva2:1013030/FULLTEXT01.pdf
```
```suggestion
[Abstract Timer]: https://ltu.diva-portal.org/smash/get/diva2:1013030/FULLTEXT01.pdf
```

AfoHT commented on book/en/src/preface.md:
```diff
@@ -10,29 +10,160 @@
+
+This section gives a background to the RTIC model. Feel free to skip to section [RTIC the model](preface.md#rtic-the-model) for a TL;DR.
+
+The RTIC framework takes the outset from real-time systems research at Luleå University of Technology (LTU) Sweden. RTIC is inspired by the concurrency model of the [Timber] language, the [RTFM-SRP] based scheduler, the [RTFM-core] language and [Abstract Timer] implementation. For a full list of related research see [TODO].
```
- [x] TODO

AfoHT commented on book/en/src/preface.md:
```diff
@@ -10,29 +10,160 @@
+
+## Stack Resource Policy based Scheduling
+
+Stack Resource Policy (SRP) based concurrency and resource management is at heart of the RTIC framework. The [SRP] model itself extends on [Priority Inheritance Protocols], and provides a set of outstanding properties for single core scheduling. To name a few:
```
```suggestion
[Stack Resource Policy (SRP)][SRP] based concurrency and resource management is at heart of the RTIC framework. The SRP model itself extends on [Priority Inheritance Protocols], and provides a set of outstanding properties for single core scheduling. To name a few:
```

AfoHT commented on book/en/src/preface.md:
```diff
@@ -10,29 +10,160 @@
+
+## RTIC the hardware accelerated real-time scheduler
+
+SRP itself is compatible both to dynamic and static priority scheduling. For the implementation of RTIC we leverage on the underlying hardware for accelerated static priority scheduling.
```
```suggestion
SRP itself is compatible with both dynamic and static priority scheduling. For the implementation of RTIC we leverage on the underlying hardware for accelerated static priority scheduling.
```

AfoHT commented on book/en/src/preface.md:
```diff
@@ -10,29 +10,160 @@
+
+The similarities are striking and it is not by chance/luck/coincidence. The hardware was cleverly designed with real-time scheduling in mind. 
+
+In order to map the SRP scheduling onto the hardware we need to have a closer look on the system ceiling (𝜫). Under SRP 𝜫 is computed as the maximum priority ceiling of the currently held resources, and will thus change dynamically during the system operation.
```
```suggestion
In order to map the SRP scheduling onto the hardware we need to take a closer look at the system ceiling (𝜫). Under SRP 𝜫 is computed as the maximum priority ceiling of the currently held resources, and will thus change dynamically during the system operation.
```

AfoHT commented on book/en/src/preface.md:
```diff
@@ -10,29 +10,160 @@
 
-Another common view from the community is that RTIC is a concurrency framework as there
-is no software kernel and that it relies on external HALs.
+For the running example, a snapshot of the ARM Cortex M [NVIC] may have the following configuration (after task `A` has been pended for execution.)
```
```suggestion
For the running example, a snapshot of the ARM Cortex M [Nested Vectored Interrupt Controller (NVIC)][NVIC] may have the following configuration (after task `A` has been pended for execution.)
```

AfoHT commented on book/en/src/preface.md:
```diff
@@ -10,29 +10,160 @@
+
+Rust `async`/`await` support is still incomplete and/or under development (e.g., there are no stable way to express `async` closures, precluding use in iterator patterns). Nevertheless, Rust `async`/`await` is production ready and covers most common use cases. 
+
+An important property is that futures are composable, thus you can await either, all, or any combination of possible futures (allowing e.g., timeouts and/or asynchronous errors to be promptly handled). For more details and examples see Section [todo].
```
- [ ] TODO

AfoHT commented on book/en/src/rtic_vs.md:
```diff
@@ -0,0 +1,31 @@
+
+Comparing RTIC to traditional a Real-Time Operating System (RTOS) is hard. Firstly, a traditional RTOS typically comes with no guarantees regarding system safety, even the most hardened kernels like the formally verified [seL4] kernel. Their claims to integrity, confidentiality, and availability regards only the kernel itself (under additional assumptions its configuration and environment). They  ...
+
+"An OS kernel, verified or not, does not automatically make a system secure. In fact, any system, no matter how secure, can be used in insecure ways."
```
```suggestion
"An OS kernel, verified or not, does not automatically make a system secure. In fact, any system, no matter how secure, can be used in insecure ways." - [seL4 FAQ][sel4faq]

[sel4faq]: https://docs.sel4.systems/projects/sel4/frequently-asked-questions.html
```

korken89 committed:
Cleanup common code and clippy fixes

AfoHT commented on book/en/src/by-example/channel.md:
````diff
@@ -0,0 +1,112 @@
+{{#include ../../../../rtic/ci/expected/async-channel.run}}
+```
+
+Also sender endpoint can be awaited. In case there the channel capacity has not been reached, `await` the sender can progress immediately, while in the case the capacity is reached, the sender is blocked until there is free space in the queue. In this way data is never lost.
````
```suggestion
Also sender endpoint can be awaited. In case the channel capacity has not yet been reached, `await`-ing the sender can progress immediately, while in the case the capacity is reached, the sender is blocked until there is free space in the queue. In this way data is never lost.
```

AfoHT commented on book/en/src/by-example/channel.md:
```diff
@@ -0,0 +1,112 @@
+
+Also sender endpoint can be awaited. In case there the channel capacity has not been reached, `await` the sender can progress immediately, while in the case the capacity is reached, the sender is blocked until there is free space in the queue. In this way data is never lost.
+
+In the below example the `CAPACITY` has been reduced to 1, forcing sender tasks to wait until the data in the channel has been received.
```
```suggestion
In the following example the `CAPACITY` has been reduced to 1, forcing sender tasks to wait until the data in the channel has been received.
```

AfoHT commented on book/en/src/by-example/channel.md:
```diff
@@ -0,0 +1,112 @@
+
+## Error handling
+
+In case all senders have been dropped `await` on an empty receiver channel results in an error. This allows to gracefully implement different types of shutdown operations.
```
```suggestion
In case all senders have been dropped `await`-ing on an empty receiver channel results in an error. This allows to gracefully implement different types of shutdown operations.
```

AfoHT replied:
https://rust-lang.github.io/async-book/03_async_await/01_chapter.html

Uses this pattern, to match `await` with the sentence

AfoHT commented on book/en/src/by-example/channel.md:
````diff
@@ -0,0 +1,112 @@
+{{#include ../../../../rtic/ci/expected/async-channel-no-sender.run}}
+```
+
+Similarly, `await` on a send channel results in an error in case the receiver has been dropped. This allows to gracefully implement application level error handling.
````
```suggestion
Similarly, `await`-ing on a send channel results in an error in case the receiver has been dropped. This allows to gracefully implement application level error handling.
```

AfoHT commented on book/en/src/preface.md:
```diff
@@ -10,29 +10,160 @@
 <!--[Russian]: ../ru/index.html-->
 
-This is the documentation of v1.0.x of RTIC; for the documentation of version
+This is the documentation of v2.0.x (pre-release) of RTIC 2.
```
```suggestion
This is the documentation for RTIC v2.x. 
```

AfoHT commented on book/en/src/SUMMARY.md:
```diff
@@ -23,13 +21,13 @@
   - [v0.4.x to v0.5.x](./migration/migration_v4.md)
   - [RTFM to RTIC](./migration/migration_rtic.md)
 - [Under the hood](./internals.md)
-  - [Cortex-M architectures](./internals/targets.md)
```
Git mishap

perlindgren replied:
Just commented out stuff that is not ready for  the pre-release.

AfoHT reviewed (commented):
Book review round 1 :+1:

AfoHT committed:
CI: Require jobs to pass CI, improve job names

AfoHT committed:
Revert accidental removal of editorial changes

perlindgren committed:
Monotonic book update

perlindgren committed:
Monotonic book

korken89 committed:
Minor book fix

perlindgren committed:
book polish

AfoHT committed:
Fix cargo README paths, disable publish for xtask

Do not publish xtask crate under any circumstance

perlindgren committed:
book remove ramfunc, remove migration

AfoHT committed:
Add alpha.0 and required Cargo fields

AfoHT committed:
Add authors to each Cargo.toml

Copy-paste the default one used for the project

AfoHT committed:
Book: Update authors

perlindgren committed:
Apply suggestions from code review

Thanks for all suggestions, awesome!

Co-authored-by: Henrik Tjäder <henrik@tjaders.com>

AfoHT committed:
CI: Fix CI, missed space

AfoHT committed:
RTIC: Examples cargo fmt

korken89 committed:
Make xtask pass, clean up rtic Cargo.toml

korken89 committed:
Simplify Systick Monotonic by integrating the TQ

perlindgren committed:
deny missing docs

korken89 committed:
Fix missing crate docs in examples

perlindgren committed:
examples polished

perlindgren committed:
book/example polish

AfoHT committed:
Update book/en/src/by-example/app.md

AfoHT committed:
Book: Touchup README and preface

AfoHT committed:
rtic-monotonics: Simplify features, default is 1kHz

Make 100 Hz or 10 kHz opt in through features,
which are meant for testing primarily.

AfoHT committed:
xtask: Clippy, optionally get cargo size output

AfoHT committed:
xtask: Allow filtering examples

AfoHT committed:
Move rtic macros to repo root, tune xtask

AfoHT committed:
xtask: Make target flag optional, default to all targets

AfoHT committed:
Upgrade to clap v4, use log and env_logger

AfoHT committed:
Improve build time with Rayon

AfoHT committed:
CI: Update to match separate rtic-macros

AfoHT committed:
xtask: Add thumbv8.base thumbv8.main

AfoHT committed:
Add other crates to workspace

AfoHT committed:
Added thumbv8 to rust-toolchain

AfoHT committed:
xtask: Add proper diff printing in case of incorrect results

AfoHT committed:
xtask: Major overhaul

Fix error printouts for examples,
Rename BuildAll -> Build
Rename Build -> ExampleBuild

Command interface changed, targets by defalt only v6 and v7

AfoHT committed:
CI: Use xtask for running QEMU examples

AfoHT committed:
CI: Use xtask for all checks

AfoHT committed:
CI: Fix macros job name

AfoHT committed:
xtask: Fixup, copy paste fail

AfoHT committed:
CI: Improve naming and styling

AfoHT committed:
xtask: For now test on ARM

AfoHT committed:
Add rtic-arbiter to workspace

AfoHT committed:
CI: Let xtask running the different targets

AfoHT committed:
CI: Cargo clippy done via xtask

AfoHT committed:
CI: Build docs for all crates

AfoHT committed:
CI: xtask handling example tests

AfoHT committed:
CI: Docs use underscore, not hyphen

AfoHT committed:
CI: Make it clearer in GHA UI which target for QEMU

AfoHT committed:
xtask: Propagate stdio/stderr, exitcodes

AfoHT committed:
CI: Make clippy fail on error

korken89 committed:
RP2040 monotonic, rename systick monotonic to not have double name

korken89 committed:
rtic-monotonics: Feature gate monotonics correctly to support multiple MCUs

korken89 committed:
Break out core specific codegen to bindings

datdenkikniet committed:
Fix spelling on "cortex_m_systick"

datdenkikniet committed:
Make embedded-hal-async dependency optional for better compatibility with HALs

Some hals implement traits for embedded-hal version `=1.0.0.alpha.<not 9>`, which is
explicitly incompatible with the version `=1.0.0.alpha.9` which embedded-hal-async
depends on. Making the dependency optional allows downstream projects to include
rtic-monotonic without requiring that all of their other libraries also implement
that specific version of embedded-hal 1.0

korken89 committed:
Make some linked list operations unsafe, and document their safety at usage

korken89 committed:
Add feature flags

korken89 committed:
Added support for adding codegen to intrrupt entry and exit (needed for RISC-V)

korken89 committed:
Fix xtask to work with backends

korken89 committed:
Update backend features to be more clear

AfoHT committed:
Book: Fix links, proofread targets and starting_a_project

korken89 committed:
Add template for bindings

AfoHT committed:
xtask: Default backend needs no argument

Propagate features for check and build

AfoHT committed:
Book: Reintroduce WIP internals targets

AfoHT committed:
CI: Use backend instead of target

AfoHT committed:
xtask: Add format and format-check

AfoHT committed:
CI: Use xtask for format checking

AfoHT committed:
CI: Fail on any warning in any local crate

AfoHT committed:
xtask: Do not add rtic features to all workspace crates

AfoHT committed:
CI: cargo doc needs a backend

AfoHT committed:
CI: Iterate over backends instead of packages

AfoHT committed:
CI: Basic tests rework

AfoHT committed:
CI: Enable cache for format

AfoHT committed:
CI: Features differ from xtask backends

AfoHT committed:
CI: Switch to lychee for linkchecking

AfoHT committed:
CI: Use precompiled tools if possible

AfoHT committed:
CI: install-action without shorthand

AfoHT committed:
Book: Remove RTIC v1 Russian translation

AfoHT committed:
CI: Output detailed link checking errors

AfoHT committed:
CI: Store and reuse API docs

AfoHT committed:
CI: Fix name of downloaded tar

AfoHT committed:
CI: Make it clear what has been linkchecked

AfoHT committed:
CI: TODO for deploy

AfoHT committed:
Book: Fix two broken links

AfoHT committed:
xtask: Add docs building

AfoHT committed:
xtask: Reorder commands

AfoHT committed:
xtask: Rename cargoarg to cargologlevel

AfoHT committed:
xtask: Cleanup check and build

AfoHT committed:
xtask: Add book building

AfoHT committed:
xtask: Add tests, clean a bit

AfoHT committed:
xtask: Split out arg parsing

AfoHT committed:
xtask: Split out cargo commands

AfoHT committed:
ci: fix backend names for tests

AfoHT committed:
ci: Temporary test: all tests

AfoHT committed:
xtask: check and clippy, test without target

korken89 committed:
Merge arbiter and channel into sync

AfoHT committed:
xtask: rtic-sync

AfoHT committed:
ci: rtic-sync

AfoHT committed:
ci: rtic-sync for docs

AfoHT committed:
ci: Fix so channel is checked

AfoHT committed:
xtask: Allow passing arguments to book and doc

AfoHT committed:
xtask: Rename: name->command, command->executable

AfoHT committed:
Book: Fail if a non-existing file is included

AfoHT committed:
Fixes for repo rename to rtic

AfoHT committed:
xtask: Book: Allow fully custom arguments

AfoHT committed:
ci: Overhaul docs and deploy

AfoHT committed:
ci: Split out the actual GHA publish job

AfoHT committed:
ci: Always parse version, cleanup deps

AfoHT committed:
ci: Use ci-success

AfoHT committed:
ci: Remove reference to rtic_channel

AfoHT committed:
ci: Building older books can't use new features

AfoHT committed:
ci: Do not need a copy, move is faster

AfoHT committed:
ci: mkbookold: Make sure paths exists for the books

AfoHT committed:
ci: local variables goes out of scope

AfoHT committed:
ci: Redirect file at the root

AfoHT committed:
rtic-common: Fix safety section formatting

AfoHT committed:
ci: Add rtic-uitest feature

Fixes the trybuild ui tests for rtic crate

AfoHT committed:
CFG: Align all crates to use hyphen

AfoHT committed:
ci: oldbooks simplify

AfoHT committed:
ci: mdbookold needs mdbook

AfoHT committed:
xtask: clippy fixes

AfoHT committed:
rtic-macros: clippy fixes

AfoHT committed:
rtic-time: clippy fixes

AfoHT committed:
ci: Add v6 uitest feature

AfoHT committed:
rtic-macros: clippy fixes

AfoHT committed:
ci: Clearer merge branch job name

AfoHT committed:
ci: mdbookold needs mergetostablebranch

AfoHT committed:
xtask: rtic-monotonics: no tests argument

AfoHT committed:
Book: Tidy up preface

AfoHT committed:
rtic-monotonics: Fix tests

AfoHT reviewed (approved):
Bors merge
